server/event_logging/event.py

```diff
@@ -0,0 +1,120 @@
+        """
+        return Event(execution=execution_id, type=Event.Type.EXECUTION_FINISHED, time=time, data={})
+
+    @staticmethod
+    def action_performed(execution_id: int, time: int, player: str, action: int, patient: int, duration_s: int):
+        """
+        Creates an event representing an action performed by a player.
+        """
+        return Event(execution=execution_id, type=Event.Type.PERFORMED_ACTION, time=time, data={
+            "action": action,
+            "player": player,
+            "patient": patient,
+            "duration_s": duration_s
+        })
+
+    @staticmethod
+    def location_take_from(execution_id: int, time: int, player: str, take_location_id: int, from_location_id: int):
+        """
+        Creates an event representing the action of taking a location from another location.
+        """
+        return Event(execution=execution_id, type=Event.Type.LOCATION_TAKE_FROM, time=time, data={
+            "player": player,
+            "take_location_id": take_location_id,
+            "from_location_id": from_location_id
+        })
+
+    @staticmethod
+    def location_leave(execution_id: int, time: int, player: str, leave_location_id: int):
+        """
+        Creates an event representing the action of leaving a location.
+        """
+        return Event(execution=execution_id, type=Event.Type.LOCATION_LEAVE, time=time, data={
+            "player": player,
+            "leave_location_id": leave_location_id
+        })
+
+    @staticmethod
+    def patient_arrive(execution_id: int, time: int, player: str, patient_id: int):
+        """
+        Creates an event representing the action of a player arriving at a patient.
+        """
+        return Event(execution=execution_id, type=Event.Type.PATIENT_ARRIVE, time=time, data={
+            "player": player,
+            "patient_id": patient_id
+        })
+
+    @staticmethod
+    def patient_leave(execution_id: int, time: int, player: str, patient_id: int):
+        """
+        Creates an event representing the action of a player leaving a patient.
+        """
+        return Event(execution=execution_id, type=Event.Type.PATIENT_LEAVE, time=time, data={
+            "player": player,
+            "patient_id": patient_id
+        })
+
+    def to_dict(self):
+        return {
+            "execution": self.execution,
+            "time": self.time,
+            "type": self.type.name,
+            "data": self.data
+        }
+
+    def to_json(self):
+        return json.dumps(self.to_dict())
+
+    @staticmethod
+    def from_logged_event(logged_event: LoggedEvent):
+        return Event(execution=logged_event.execution, type=Event.Type[logged_event.type], time=logged_event.time, data=logged_event.data)
+
+    @staticmethod
+    def from_json(json_str: str):
+        data = json.loads(json_str)
+        return Event(execution=data["execution"], type=Event.Type[data["type"]], time=data["time"], data=data["data"])
+
+    def log(self):
```

Comment:
Logging the event to the DB should happen as soon as the event is created, i.e. in the constructor

Reply:
I get you idea, however to enforce the log method on every event creation would make the implementation more complicated. This way there might  be an error case of creating but not logging the event. I would advocate the "shit in - shit out" technique. The Event class only provides you with possibilities.